### PR TITLE
Apps layer: Add WAW adjacency matrix rescaling function

### PR DESCRIPTION
**Pull request by trbromley**
This adds a `waw_matrix` function that rescales in input adjacency matrix according to the WAW encoding, with the objective of embedding node weights into the adjacency matrix.

### CODE REVIEW

File: tests/apps/test_sample.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 adj_dim_range = range(2, 6)
 
-t0 = np.array([0., 0., 0., 0., 0., 0., 0.])
+t0 = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
```
Comment (trbromley) [inline tests/apps/test_sample.py]: (result of running `black`)

File: strawberryfields/apps/sample.py [hunk 2]
```diff
@@ -361,3 +377,67 @@ def vibronic(
...
+    dim = len(A)
+
+    w_check = (len(w_s) == 1 and w_s[0] == dim) or (
+        len(w_s) == 2 and sorted(np.unique(w_s)) == [1, dim]
+    )  # Check if the vector has one dimension or is a two-dimensional row or column vector
+
+    if not w_check:
+        raise ValueError("Vector of node weights must be a row or column vector")
+
+    return w.flatten()
```
Comment (trbromley) [inline strawberryfields/apps/sample.py]: I plan to use this function for input validation of the weight vector.

We definitely want to let users input either a list or numpy array without throwing an error.
This function also checks the shape of the numpy array.

One question - where is the best place for this function to live? It will be used in other modules (e.g. `clique`). I think it's perhaps ok for it to just live here for now, it is not a user-facing function and is not shown in the documentation.

File: strawberryfields/apps/sample.py [hunk 3]
```diff
@@ -86,9 +86,25 @@
 
 The subgraphs sampled from GBS are likely to be dense :cite:`arrazola2018using`, motivating their
 use within heuristics for problems such as maximum clique (see :mod:`~.apps.clique`).
+
+Including node weights
+----------------------
+
+Some graphs are composed of nodes with weights. These node weights can correspond to relevant
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]:
```suggestion
Some graphs are composed of nodes with weights. These weights can correspond to relevant
```

File: strawberryfields/apps/sample.py [hunk 4]
```diff
@@ -86,9 +86,25 @@
...
+information in an optimization problem and it is desirable to encode them into GBS along with the
+graph's adjacency matrix. One canonical approach to doing this is to use the ``WAW`` encoding
+:cite:`banchi2019molecular`, which rescales the adjacency matrix according to:
+
+.. math::
+
+    A \rightarrow WAW,
+
+with :math:`W` the diagonal matrix formed by the weighted nodes. The rescaled adjacency matrix
+can be passed to :func:`sample`, resulting in a bias toward sampling subgraphs with a large
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: I would rephrase this as: 
can be passed to sample, resulting in a distribution that increases the probability of observing a node proportionally to its weight.

File: strawberryfields/apps/sample.py [hunk 5]
```diff
@@ -86,9 +86,25 @@
...
+Some graphs are composed of nodes with weights. These node weights can correspond to relevant
+information in an optimization problem and it is desirable to encode them into GBS along with the
+graph's adjacency matrix. One canonical approach to doing this is to use the ``WAW`` encoding
+:cite:`banchi2019molecular`, which rescales the adjacency matrix according to:
+
+.. math::
+
+    A \rightarrow WAW,
+
+with :math:`W` the diagonal matrix formed by the weighted nodes. The rescaled adjacency matrix
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: It may be good to write W explicitly. One compact way of doing so is using `W=w_{ij}\delta_{ij}`

File: strawberryfields/apps/sample.py [hunk 6]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+    r"""Rescale adjacency matrix to account for node weights.
+
+    Given a graph with adjacency matrix :math:`A` and a vector :math:`w` of weighted nodes,
+    this function rescales the adjacency matrix according to:
+
+    .. math::
+
+        A \rightarrow WAW,
+
+    with :math:`W` the diagonal matrix formed by the weighted nodes. This encoding can be used to
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: Again maybe good to write W explicitly and to rephrase the effect on the distribution

File: strawberryfields/apps/sample.py [hunk 7]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+    :cite:`banchi2019molecular`. The resulting matrix can be passed to :func:`sample`.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(5, 0.7)
+    >>> a = nx.to_numpy_array(g)
+    >>> w = [1, 1, 3, 1, 0.5]
+    >>> a = waw_matrix(a, w)
+    >>> sample(a, 3, 4)
+    [[0, 0, 0, 0, 0], [1, 0, 0, 0, 1], [0, 0, 0, 0, 0], [0, 1, 1, 1, 1]]
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: Can we `post-select` on samples that more clearly show that nodes with large weight appear more often?

File: strawberryfields/apps/sample.py [hunk 8]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+
+    with :math:`W` the diagonal matrix formed by the weighted nodes. This encoding can be used to
+    add a bias in GBS toward selecting subgraphs with a large product of node weights
+    :cite:`banchi2019molecular`. The resulting matrix can be passed to :func:`sample`.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(5, 0.7)
+    >>> a = nx.to_numpy_array(g)
+    >>> w = [1, 1, 3, 1, 0.5]
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: Maybe `w=[1,2,3,4,5]` so that it's more clear which nodes have larger weight?

File: strawberryfields/apps/sample.py [hunk 9]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+
+    Returns:
+        array: matrix rescaled according to the WAW encoding
+    """
+    if not np.allclose(A, A.T):
+        raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+
+    w = _process_w(len(A), w)
+
+    return (w * A).T * w
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: Clever!

File: strawberryfields/apps/sample.py [hunk 10]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+
+
+def _process_w(dim: int, w: Union[np.ndarray, list]) -> np.ndarray:
```
Comment (ixfoduap) [inline strawberryfields/apps/sample.py]: I don't mind this function, but isn't it a bit of overkill?

Comment (ixfoduap) [reply strawberryfields/apps/sample.py]: Is the idea that we are allowing the weight vector to have different forms? Isn't it better to have a fixed form?

File: strawberryfields/apps/sample.py (hunk 3 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Thanks!

File: strawberryfields/apps/sample.py (hunk 5 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Nice! Done

File: strawberryfields/apps/sample.py (hunk 4 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Thanks have added

File: strawberryfields/apps/sample.py (hunk 6 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Done

File: strawberryfields/apps/sample.py (hunk 10 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: The user might want to input the weight vector as:
- a list
- a numpy array of shape (dim,) - standard
- a numpy array of shape (1, dim) - a 2d row vector
- a numpy array of shape (dim, 1) - a 2d column vector

Actually the above line 417 is fine with any of them... so maybe you're right and we can skip the checking. Although one thing is we definitely should let the user input as a list or numpy array.

Comment (trbromley) [reply strawberryfields/apps/sample.py]: Ok, simplified

File: strawberryfields/apps/sample.py (hunk 8 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: I like having it be a bit different to consecutive ascending numbers, as maybe `[1,2,3,4,5]` could confuse someone that the weights have to be linked to the node numbers?

File: strawberryfields/apps/sample.py (hunk 7 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]:
```rst
>>> g = nx.erdos_renyi_graph(5, 0.7)
>>> a = nx.to_numpy_array(g)
>>> w = [10, 1, 0, 1, 1]
>>> a = waw_matrix(a, w)
>>> sample(a, 3, 4)
[[1, 1, 0, 1, 0], [1, 1, 0, 1, 0], [1, 1, 0, 0, 0], [1, 1, 0, 1, 0]]
```
This is what I came up with. There is a zero weight one which nicely never clicks. There is a 10 weight one which always clicks

File: strawberryfields/apps/sample.py [hunk 11]
```diff
@@ -361,3 +377,66 @@ def vibronic(
...
+    with :math:`W` the diagonal matrix formed by the weighted nodes. This encoding can be used to
+    add a bias in GBS toward selecting subgraphs with a large product of node weights
+    :cite:`banchi2019molecular`. The resulting matrix can be passed to :func:`sample`.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(5, 0.7)
+    >>> a = nx.to_numpy_array(g)
+    >>> w = [1, 1, 3, 1, 0.5]
+    >>> a = waw_matrix(a, w)
```
Comment (soranjh) [inline strawberryfields/apps/sample.py]: It might be nice to print the new a here.

File: strawberryfields/apps/sample.py (hunk 7 above)
Comment (soranjh) [reply strawberryfields/apps/sample.py]: Or maybe use large weights such that all samples reflect that without post selecting?

File: tests/apps/test_sample.py [hunk 12]
```diff
@@ -368,3 +369,48 @@ def test_vibronic_integration(p, integration_sample_number):
...
+    @pytest.mark.parametrize("inst", zip(adjs, wvecs, resc_adjs))
+    def test_valid_w(self, inst):
+        """Test if function returns the correct answer on some pre-calculated instances."""
+        assert np.allclose(sample.waw_matrix(inst[0], inst[1]), inst[2])
+
+    @pytest.mark.parametrize("inst", zip(adjs, wvecs, resc_adjs))
+    def test_valid_w_list(self, inst):
+        """Test if function returns the correct answer on some pre-calculated instances,
+        when w is a list"""
+        assert np.allclose(sample.waw_matrix(inst[0], list(inst[1])), inst[2])
```
Comment (josh146) [inline tests/apps/test_sample.py]: Nice!

File: strawberryfields/apps/sample.py [hunk 13]
```diff
@@ -361,3 +377,42 @@ def vibronic(
...
+    Returns:
+        array: matrix rescaled according to the WAW encoding
+    """
+    if not np.allclose(A, A.T):
+        raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+
+    if isinstance(w, list):
+        w = np.array(w)
+
+    return (w * A).T * w
```
Comment (josh146) [inline strawberryfields/apps/sample.py]: Maybe I'm missing something (because the tests below pass), but is this doing WAW? I did a small test in NumPy:

```python
>>> import numpy as np
>>> A = np.random.random(size=[4,4])
>>> w = np.random.random(size=[4])
>>> (w * A).T * w == np.diag(w) @ A @ np.diag(w)
array([[ True, False, False, False],
       [False,  True, False, False],
       [False, False,  True, False],
       [False, False, False,  True]])
```

Comment (josh146) [reply strawberryfields/apps/sample.py]: It seems that the diagonal is rescaled correctly, but the off-diagonals differ. But let me know if I'm doing something silly 😆

File: tests/apps/test_sample.py [hunk 14]
```diff
@@ -368,3 +369,48 @@ def test_vibronic_integration(p, integration_sample_number):
...
+            ]
+        ),
+        np.array([[0, -2, 0, -4], [-2, 0, -6, 0], [0, -6, 0, -12], [-4, 0, -12, 0]]),
+        np.array([[0, 0, 0], [0, 0, 1], [0, 1, 0]]),
+    ]
+
+    def test_invalid_adjacency(self):
+        """Test if function raises a ``ValueError`` for a matrix that is not symmetric."""
+        with pytest.raises(ValueError, match="Input must be a NumPy array"):
+            adj_asym = np.triu(np.ones((4, 4)))
```
Comment (josh146) [inline tests/apps/test_sample.py]: You could place this line before the pytest.raises, since it will not raise the ValueError (just makes it slightly cleaner)

File: tests/apps/test_sample.py [hunk 15]
```diff
@@ -368,3 +369,48 @@ def test_vibronic_integration(p, integration_sample_number):
...
+            sample.sample(A=adj_asym, n_mean=1.0)
+
+    @pytest.mark.parametrize("inst", zip(adjs, wvecs, resc_adjs))
+    def test_valid_w(self, inst):
+        """Test if function returns the correct answer on some pre-calculated instances."""
+        assert np.allclose(sample.waw_matrix(inst[0], inst[1]), inst[2])
```
Comment (josh146) [inline tests/apps/test_sample.py]: The passing tests make me think that my comment above about the WAW calculation is incorrect :)

File: tests/apps/test_sample.py (hunk 12 above)
Comment (josh146) [reply tests/apps/test_sample.py]: How were the matrices pre-calculated? By hand?

File: strawberryfields/apps/sample.py [hunk 16]
```diff
@@ -86,9 +86,25 @@
 
 The subgraphs sampled from GBS are likely to be dense :cite:`arrazola2018using`, motivating their
 use within heuristics for problems such as maximum clique (see :mod:`~.apps.clique`).
+
+Including node weights
+----------------------
+
+Some graphs are composed of nodes with weights. These weights can correspond to relevant
+information in an optimization problem and it is desirable to encode them into GBS along with the
+graph's adjacency matrix. One canonical approach to doing this is to use the ``WAW`` encoding
```
[the remark this answers is not in the document]
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Nice, fixed to math mode.

As a side note... I keep accidentally writing WAR instead of WAW :thinking:

File: strawberryfields/apps/sample.py (hunk 11 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: That's a great idea, have done this. Makes the code block a bit longer but I think it's worth it.
image

File: strawberryfields/apps/sample.py (hunk 13 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Your code block looks right, but in this case A is symmetric. We have:
`w * A =  A @ np.diag(w)`
`(w * A).T = np.diag(w) @ A.T`
`(w * A).T = np.diag(w) @ A`
`(w * A).T * w = np.diag(w) @ A.T @ np.diag(w)`
So the `(w * A).T * w` approach does require symmetry of `A`, and wouldn't work if `A` was complex and hermitian. However, I think we are reasonably restricting to graphs with real edge weights for now.

Although, perhaps going with `np.diag(w) @ A @ np.diag(w)` is simpler. One minor drawback of `np.diag(w) @ A @ np.diag(w)` is that it doesn't work vectors that are 2D arrays (e.g. row or column vector), while the `(w * A).T * w` approach does - but this setting should be relatively rare.

File: tests/apps/test_sample.py (hunk 14 above)
Comment (trbromley) [reply tests/apps/test_sample.py]: Nice spot, done!

File: tests/apps/test_sample.py (hunk 12 above)
Comment (trbromley) [reply tests/apps/test_sample.py]: Yes all by :hand:
Tried to cover a variety of cases (e.g. edges with different weights in `A`, positive/negative/fractional node weights)

File: tests/apps/test_sample.py (hunk 15 above)
Comment (trbromley) [reply tests/apps/test_sample.py]: Yes, due to the symmetry of the test `A`'s

File: strawberryfields/apps/sample.py (hunk 7 above)
Comment (trbromley) [reply strawberryfields/apps/sample.py]: Thanks, yes we're now using some more biased weights and picking some indicative samples.
